mmdet/utils/memory.py

```diff
@@ -0,0 +1,172 @@
+        error. It will do the following steps:
+
+            1. first retry after calling `torch.cuda.empty_cache()`.
+            2. If that still fails, it will then retry by trying to
+              convert inputs to FP16.
+            3. If that still fails trying to convert inputs to CPUs.
+              In this case, it expects the function to dispatch to
+              CPU implementation.
+
+        Args:
+            func: a stateless callable that takes tensor-like objects
+                as arguments
+        Returns:
+            func: a callable which retries `func` if OOM is encountered.
+        """
+
+        def maybe_to_cpu(x):
+            try:
+                like_gpu_tensor = x.device.type == 'cuda' and hasattr(x, 'to')
+            except AttributeError:
+                like_gpu_tensor = False
+            if like_gpu_tensor:
+                return x.to(device='cpu')
+            else:
+                return x
+
+        def maybe_to_fp16(x):
+            try:
+                like_float_tensor = \
+                    x.dtype == torch.float32 or x.dtype == torch.float64
+            except AttributeError:
+                like_float_tensor = False
+            if like_float_tensor:
+                return x.to(torch.half)
+            else:
+                return x
+
+        @wraps(func)
+        def wrapped(*args, **kwargs):
+            if self.logger is None:
+                self.logger = get_root_logger()
+
+            # raw function
+            with _ignore_torch_cuda_oom():
+                return func(*args, **kwargs)
+
+            # Clear cache and retry
+            torch.cuda.empty_cache()
+            with _ignore_torch_cuda_oom():
+                return func(*args, **kwargs)
+
+            # Try to use FP16
+            fp16_args = (maybe_to_fp16(x) for x in args)
+            fp16_kwargs = {k: maybe_to_fp16(v) for k, v in kwargs.items()}
+            self.logger.info(f'Attempting to copy inputs of {str(func)} '
+                             f'to fp16 due to CUDA OOM')
+            if self.out_fp32:
+                if self.output_type is not None:
+                    dtype = self.output_type
+                else:
+                    dtype = torch.float
+                with _ignore_torch_cuda_oom():
+                    self.logger.info(f'Trying to convert output to {dtype}')
+                    return func(*fp16_args, **fp16_kwargs).to(dtype=dtype)
+                self.logger.info(f'Cannot convert output to {dtype} due to '
+                                 'CUDA OOM.')
+            with _ignore_torch_cuda_oom():
+                # try to convert outputs to fp16
+                self.logger.info('Trying to convert outputs to fp16')
+                return func(*fp16_args, **fp16_kwargs)
+            warnings.warn('Cannot convert outputs to fp16 due to CUDA OOM')
+            self.logger.info('Cannot convert output to fp16 due to '
+                             'CUDA OOM.')
```

Comment:
Repeat implement `func` with fp16, we can merge this branch into above `if self.out_fp32` branch.